### PR TITLE
[MRG] BUG: don't normalize when confounds

### LINKED ISSUE #549: signal.clean with standardize=False
The behavior of signal.clean with standardize=False is different whether one includes or not confounds. If confounds to remove, the input signal and the confounds are standardized even though standardize=False.

The standard choice for scaling signals within ROIs (e.g. MarsBaR, Conn) is to scale them to their own ROI-specific average BOLD signal and multiply by 100 to get what is commonly called Percent Signal Change units. This allows to put all the ROIs to the same range while maintaining the difference in variances between them. 

Below signals within 2 ROIs from ADHD dataset, note that although region 12 has higher average, it has less variance. But after the regression of one confound (average CSF), the variances are almost the same because of this partial standardization achieved in signal.clean. My expected behavior of cleaning is the bottom subplot.

demo_psc

### PR DESCRIPTION
**Pull request by GaelVaroquaux**
Fixes #549

The signals were normlized when confounds were used, for little good
reason

### CODE REVIEW

File: nilearn/signal.py [hunk 1]
```diff
@@ -26,20 +26,24 @@ def _standardize(signals, detrend=False, normalize=True):
 
     Parameters
     ==========
-    signals: numpy.ndarray
+    signals : numpy.ndarray
```
Comment (AlexandreAbraham) [inline nilearn/signal.py]: Do we have to put space before colons now?

Comment (GaelVaroquaux) [inline nilearn/signal.py]: > Do we have to put space before colons now?

Yes. It's always been like this in the numpy doc standard. But we keep
forgetting.

Comment (AlexandreAbraham) [inline nilearn/signal.py]: Well, I don't keep forgetting, I've been corrected several times when I add them...

Comment (GaelVaroquaux) [inline nilearn/signal.py]: > Well, I don't keep forgetting, I've been corrected several times when I add
> them...

:$

Comment (AlexandreAbraham) [inline nilearn/signal.py]: Here is an image of generated doc. For the first parameter, there is no space. For the second one, I put a space. You can see that the extra space shows in the doc. This is, AFAIK, why we decided not to put them (this is the same in scikit-learn). I am not against adding them but we should be consistent and do that in a dedicated PR.

space

File: nilearn/signal.py [hunk 2]
```diff
@@ -26,20 +26,24 @@ def _standardize(signals, detrend=False, normalize=True):
...
 
     Returns
     =======
-    std_signals: numpy.ndarray
+    std_signals : numpy.ndarray
         copy of signals, normalized.
+
+    std : numpy.ndarray
+        The standard deviation of the signals: multiply the output
+        signals by this quantity to get the original signals back
```
Comment (AlexandreAbraham) [inline nilearn/signal.py]: There is no change in the code for that output.

Comment (salma1601) [inline nilearn/signal.py]: hi @GaelVaroquaux 
What's the need for this additionnal output std ? Can't we just not standardize the input signal in the confounds removal if standardize is set to False?

Comment (GaelVaroquaux) [inline nilearn/signal.py]: Indeed, thanks for the comment, Alex and Salma. This is a mistake and I
am correcting it.

**Comment by GaelVaroquaux** [comment]
Comments addressed. Thanks for the review.

File: nilearn/signal.py [hunk 3]
```diff
@@ -408,14 +408,9 @@ def clean(signals, detrend=True, standardize=True, confounds=None,
...
-    # Standardize / detrend
-    normalize = False
-    if confounds is not None:
-        # If confounds are to be removed, then force normalization to improve
-        # matrix conditioning.
-        normalize = True
+    # detrend
     signals = _ensure_float(signals)
-    signals = _standardize(signals, normalize=normalize, detrend=detrend)
+    signals = _standardize(signals, normalize=False, detrend=detrend)
```
Comment (salma1601) [inline nilearn/signal.py]: if detrend is False here, the _standardize will just return a copy of the signals without centering them. May be the function _standardize also has to be changed so that the signal is centered even if both standardize and detrend are False ?

Comment (AlexandreAbraham) [inline nilearn/signal.py]: You mean that you want a constant trend to be removed even if detrend is False?

Comment (salma1601) [inline nilearn/signal.py]: yes that's what I meant but thinking again it is not necessary. In fact I was testing the new code on ADHD removing one unique constant confound and I had a weired result.
demo_constant_removal

Investigating what is wrong it turns out it is the way the regression of confounds is done. In fact, it relies on a QR decomposition and in case of a unique null confound, the Q matrix is the vector np.array([1, 0, ..., 0]). So the output signal is np.array([0, input[1], ..., input[n]]). 

https://github.com/nilearn/nilearn/blob/master/nilearn/signal.py#L460

I think that the case of a constant confound must be handled precisely. Shall I open another issue for that?

Comment (salma1601) [inline nilearn/signal.py]: I added a new issue for the confounds removal, because I think it is unrelated to standardization.
https://github.com/nilearn/nilearn/issues/561

**Comment by GaelVaroquaux** [comment]
By looking at the discussion, I think that they is nothing to do on this PR, and that it is mergeable.

Can I haz reviews? To haz merge?

File: nilearn/tests/test_signal.py [hunk 4]
```diff
@@ -279,7 +279,7 @@ def test_clean_confounds():
     # With signal: output must be orthogonal to confounds
     cleaned_signals = nisignal.clean(signals + noises, confounds=confounds,
                                       detrend=True, standardize=False)
-    assert_true(abs(np.dot(confounds.T, cleaned_signals)).max() < 100. * eps)
+    assert_true(abs(np.dot(confounds.T, cleaned_signals)).max() < 1000. * eps)
```
Comment (lesteve) [inline nilearn/tests/test_signal.py]: I feel this has been discussed but I couldn't find it anywhere. How come we need to relax this condition, i.e. somehow that the cleaned_signals and confounds are less orthogonal ?

File: nilearn/tests/test_signal.py [hunk 5]
```diff
@@ -296,7 +296,7 @@ def test_clean_confounds():
                                       detrend=True, standardize=False)
     coeffs = np.polyfit(np.arange(cleaned_signals.shape[0]),
                         cleaned_signals, 1)
-    assert_true((abs(coeffs) < 10. * eps).all())  # trend removed
+    assert_true((abs(coeffs) < 100. * eps).all())  # trend removed
```
Comment (lesteve) [inline nilearn/tests/test_signal.py]: Same question as above, do we understand why we need to relax this condition?

File: nilearn/tests/test_signal.py (hunk 4 above)
Comment (GaelVaroquaux) [inline nilearn/tests/test_signal.py]: I am not sure. I couldn't get this to work.

There are two possible explanations: either the normalizing does indeed improve the numerics, and the results are less orthogonal. I think that it would not be the end of the world given that the precision is still very good.

Or options: the norm of residuals is really related to the norm of what goes in. Relative errors are what it meaningful, not absolute. Thus, it is normal that it goes up.

Anyhow, I couldn't decide which one was true, and chose to move forward: I don't think that this is too bad.

File: nilearn/tests/test_signal.py (hunk 5 above)
Comment (salma1601) [inline nilearn/tests/test_signal.py]: >Or options: the norm of residuals is really related to the norm of what goes in. Relative errors are what it meaningful, not absolute. Thus, it is normal that it goes up.

for me this is the reason we need to relax both tests. To be convinced, rerun the same tests on the master with standardize=True (which results in a multiplication of the output signal by some factor) and the tests will not work.

**Comment by AlexandreAbraham** [comment]
Pouce vers le haut !

**Comment by GaelVaroquaux** [comment]
Okay! Alors je merge ce tantôt.

**Comment by eickenberg** [comment]
> Pouce vers le haut !

Bon pour fusion?
